### PR TITLE
[FLOC-3767] Container volume support in API client

### PR DESCRIPTION
**Pull request by jongiddy**
For benchmarking, we still require the container API. However, the `flocker.apiclient.Client` does not support volumes attached to containers, which is needed for benchmarking.

This PR adds support for stateful containers to the API client, both when creating containers, and when receiving information about container configuration and state.

Note the use of PEP 484 type annotations in the docstrings. Currently we have a variety of patterns for non-primitive docstring types (including `|`, top-level type only (`dict`), verbose descriptions, and leaving them out completely).  Despite there being no implementation support for them in Python 2, the PEP 484 type annotations work particularly well for a function like `_parse_volumes`, where they quickly make apparent the transformation taking place.

### CODE REVIEW

File: flocker/apiclient/_client.py [hunk 1]
```diff
@@ -760,15 +801,23 @@ def _parse_configuration_container(self, container_dict):
         :return: ``Container`` instance.
         """
         return Container(
-            node_uuid=UUID(hex=container_dict[u"node_uuid"], version=4),
+            node_uuid=UUID(hex=container_dict[u"node_uuid"]),
```
Comment (jongiddy) [inline flocker/apiclient/_client.py]: This change is not really related to this PR. But setting the version of a UUID only makes sense if we are minting a new UUID from some data. Overriding the version when creating a UUID instance from existing UUID data is asking for trouble.

File: flocker/apiclient/_client.py [hunk 2]
```diff
@@ -111,34 +111,51 @@ class Lease(PClass):
...
     :attr DockerImage image: The name of the Docker image.
     :attr bool running: Whether the container is running.
+    :attr Optional[Sequence[MountedDataset]] volumes: Flocker volumes
+        mounted in container.
     """
     node_uuid = field(type=UUID, mandatory=True)
     name = field(type=unicode, mandatory=True)
     image = field(type=DockerImage, mandatory=True)
     running = field(type=bool, mandatory=True)
+    volumes = field(initial=None)
```
Comment (jml) [inline flocker/apiclient/_client.py]: Missing `mandatory=True` means that some instances of this might not have the `volumes` attribute set at all (i.e. `state.volumes` will raise an `AttributeError`). Is this deliberate?

File: flocker/apiclient/_client.py [hunk 3]
```diff
@@ -751,6 +772,26 @@ def version(self):
             b"GET", b"/version", None, {OK}
         )
 
+    @staticmethod
+    def _parse_volumes(volume_list):
```
Comment (jml) [inline flocker/apiclient/_client.py]: :question: I'd make this a function, since there's no reason for inheritance.

File: flocker/apiclient/_client.py (hunk 1 above)
Comment (jml) [inline flocker/apiclient/_client.py]: Thanks.

File: flocker/apiclient/_client.py [hunk 4]
```diff
@@ -751,6 +772,26 @@ def version(self):
...
+        :return Optional[Sequence[MountedDataset]]: List of mounted datasets,
+            or None if no volumes.
+        """
+        if volume_list:
+            return [
+                MountedDataset(
+                    dataset_id=UUID(volume[u'dataset_id']),
+                    mountpoint=volume[u'mountpoint'],
+                ) for volume in volume_list
+            ]
```
Comment (jml) [inline flocker/apiclient/_client.py]: :question: Maybe make this an generator expression?

**Comment by jml** [comment]
I'm happy w/ the PEP 484 things in docstrings.

Thanks!

File: flocker/apiclient/_client.py (hunk 2 above)
Comment (jongiddy) [inline flocker/apiclient/_client.py]: @jml I guess I'm confused by the relationship between `initial` and `mandatory`. It seems to me that `initial` implies `mandatory`. Is it possible to create a PClass with a missing field if the field is defined with an `initial` value?

Even removing an attribute as an evolution seems to return it to the initial value:
```python
class X(PClass):
    a = field(initial=1)

x = X()
x = x.set(a=2)
x = x.remove('a')
assert x.a == 1
```

Comment (jml) [inline flocker/apiclient/_client.py]: OK, it looks like I was confused. Sorry.

Comment (jongiddy) [inline flocker/apiclient/_client.py]: No worries. PRecord works that way, and it does seem reasonable that removing an attribute should either completely remove the attribute or fail if the attribute is mandatory.  To the point that the existing behaviour of PClass might be a bug.  However, fixing the "bug" would be a major backwards-incompatible change.

Comment (jongiddy) [inline flocker/apiclient/_client.py]: I have filed https://github.com/tobgu/pyrsistent/issues/76

Comment (jongiddy) [inline flocker/apiclient/_client.py]: I've added the `mandatory` property, in case Pyrsistent (which is pre-1.0) changes in this regard.

File: flocker/apiclient/_client.py (hunk 4 above)
Comment (jongiddy) [inline flocker/apiclient/_client.py]: Generator expressions are iterators, but the return value must be a restartable iterable.
